### PR TITLE
Feature/initial v2

### PR DESCRIPTION
**Pull request by whaeck**
A preliminary prototype implementation of NDItk for reading, writing and modifying the multigroup neutron and photon NDI tables.

The capability is limited but it illustrates how most of the toolkit will work. We cover the basic cases that we should encounter while reading records:
  - reading metadata
  - reading simple records consisting of lists of reals and integers
  - reading more complex records that need to be divided into subrecords
  - basic verification of the data to check consistency (e.g. are the number of groups consistent across the table)

The python bindings are currently missing, but I will add these shortly.

### CODE REVIEW

**Review by tgsaller — COMMENTED** [review]
I'll look at this a lot more next week, but here are some scattered comments for now.

File: src/NDItk/multigroup/Metadata.hpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+#include "NDItk/base/SingleIntegerRecord.hpp"
+#include "NDItk/base/SingleRealRecord.hpp"
+#include "NDItk/base/SingleStringRecord.hpp"
+
+namespace njoy {
+namespace NDItk {
+namespace multigroup {
+
+/**
+ *  @brief Metadata associated to a neutron NDI file
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/Metadata.hpp:17]: Consider it changed (I'll push a commit with changes later).

As a side note: we should not shy away from adding more rather than less documentation. If you think we can elaborate, we can add more documentation in a future PR.

File: src/NDItk/multigroup/Metadata.hpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+ */
+class Metadata {
+
+  /* fields */
+
+  base::SingleStringRecord zaid_;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/Metadata.hpp:23]: The underscore is simply a coding style convention to identify private member variables. You'll sometimes see the underscore in front of the variable name, but we do not use that style since prepending a variable name with underscores is common practice for standard library development.

File: src/NDItk/base/IntegerListRecord.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  /* constructor */
+
+  /**
+   *  @brief Constructor
+   *
+   *  @param[in] keyword   the keyword of the record
+   */
+  IntegerListRecord( std::string keyword ) :
+      ListRecord( std::move( keyword ) ) {}
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/IntegerListRecord.hpp:39]: The argument for the constructor here is an std::string (not a const std::string& or a string&&) meaning that this keyword is already a local copy. As a result, we can use move semantics to simply move the underlying memory around.

Defining a constructor this way allows a user to write code with varying intents:
```
std::string key = "test";
IntegerListRecord foo( key ); // this copies key and leaves key intact after the constructor call
IntegerListRecord bar( std::move( key ) ); // this moves the underlying memory from key into the constructor call
```

This is a very common way of doing things in C++ right now.

In the unit tests, you'll often see me use this second version. When as a user knows the data is local, you can just move it. ENDFtk does not use this approach but ACEtk does (I might go back to ENDFtk and change it there as well).

File: src/NDItk/base/Record.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   *  @brief Print the record keyword
+   *
+   *  @param[in] iter   the current position in the output
+   */
+  template< typename OutputIterator >
+  void print( OutputIterator& iter ) const {
+
+    for ( auto c : this->keyword() ) { *iter++ = c; }
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/Record.hpp:40]: These iterators point to a range of characters (typically a string) so we can only do one character at a time. This is the way it was set up in ENDFtk so we continue to do it this way in ACEtk and now NDItk to keep consistency.

File: CMakeLists.txt [hunk 5]
```diff
@@ -33,8 +33,11 @@ cmake_dependent_option(
 cmake_dependent_option(
   NDItk.python
   "Build NDItk python bindings" ON
-  "NOT ${subproject}" OFF
+  "NOT ${subproject} OR DEFINED requires.NDItk.python " OFF
 )
+if ( NDItk.python )
```
[the remark this answers is not in the document]
Comment (whaeck) [reply CMakeLists.txt:38]: This is actually switched on by default. The library is already set up to do the python bindings and a dynamic python library is actually compiled when doing `make`, it's just has no code in it for now.

File: cmake/release_dependencies.cmake [hunk 6]
```diff
@@ -10,6 +10,11 @@ FetchContent_Declare( Catch2
     GIT_TAG         3f0283de7a9c43200033da996ff9093be3ac84dc # tag: v3.3.2
     )
 
+FetchContent_Declare( fast_float
+    GIT_REPOSITORY  https://github.com/fastfloat/fast_float
```
[the remark this answers is not in the document]
Comment (whaeck) [reply cmake/release_dependencies.cmake:14]: Yes, offline builds are currently an issue. We will soon be moving to an updated cmake setup to better integrate with other LANL projects that will solve this offline build issue. ACEtk will be the first to get this treatment, after which we will roll this out to the other libraries.

File: src/NDItk/MultigroupTable.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   *  @brief Return the metadata of the table
+   */
+  const multigroup::Metadata& metadata() const { return this->metadata_; }
+
+  /**
+   *  @brief Return the primary group structure record
+   */
+  const multigroup::Structure& structure() const { return this->structure_; }
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable.hpp:49]: C++ still has structs ;-)

Consider it changed. This is a major part of what reviews are for: I need someone to look at the API and be like "this could be better" :-)

[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable.hpp:49]: As a related question: should we change Structure to GroupStructure too (similar to FluxWeights)?

File: src/NDItk/MultigroupTable.hpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+   *  @param[in] iter   the current position in the output
+   */
+  template< typename OutputIterator >
+  void print( OutputIterator& iter ) const {
+
+    this->metadata_.print( iter );
+    this->structure_.print( iter );
+    this->weights_.print( iter );
+    this->xs_.print( iter );
+    *iter++ = 'e';
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable.hpp:78]: The iterator points to a single character, so yes. It's a bit clunky but it is not much of a concern.

File: src/NDItk/multigroup/ReactionCrossSections/src/generateBlocks.hpp [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+void generateBlocks() {
+
+  if ( ! this->empty() ) {
+
+    this->xs_.clear();
+    for ( unsigned int index = 0; index < this->numberReactions(); ++index ) {
+
+      const auto left = this->values().begin() + index * ( 2 + this->numberGroups() );
+      const auto right = left + 2 + this->numberGroups();
+      this->xs_.emplace_back( left, right );
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/ReactionCrossSections/src/generateBlocks.hpp:13]: Yes, this part of the code chunks up the record into a number of subrecords defined by a begin and end iterator over the full record array. emplace_back() is like push_back(), except that the underlying object (in this case the CrossSection subrecord) is created in place.

CrossSection itself is a subrecord that takes care of the abstraction of the subrecord - as you will see.

File: src/NDItk/base/Record.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/Record.hpp:40]: The iterator typically points to a string, but we can define iterators to file streams too.

File: src/NDItk/MultigroupTable.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tgsaller) [reply src/NDItk/MultigroupTable.hpp:49]: Yes.

[the remark this answers is not in the document]
Comment (joewpeterson) [reply src/NDItk/MultigroupTable.hpp:49]: Work style question: are we marking comments as resolve prior to them being implemented? Should we not push an update to the branch with the PR prior to marking this a resolved?

File: src/NDItk/base/Record.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (joewpeterson) [reply src/NDItk/base/Record.hpp:40]: This is pretty confusing to me also. The name of the function is print. I don't think see how that is a print. Perhaps a name change is in order.

There may be internal logic which makes this safe, but as an abstract thing (which it is as a base class), it strikes me as dangerous. Consider:
If record.keyword is a string of size n and iter point to a string, inString, of of size m, what are we doing. We're copying record.keyword into inString and then adding a '\n' after all copying is done, yes? What if m <= n? The null terminating char of inString is replaced by \n in the case of m=n, yes? for m<n we start accessing memory which is out of bounds

File: src/NDItk/MultigroupTable.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable.hpp:49]: Depends, if it is the only outstanding PR then we just make the modifications as additional commits to the PR. In this case, this is already 3 PRs behind (we have had even worse). For cases like these, we often use a separate update branch or refer to other PRs where things were changed (or where the issue became obsolete).

File: src/NDItk/base/Record.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/Record.hpp:40]: I was not as clear as I should have been. If you use a normal iterator (such as iter = string.begin()), then yes: you have to ensure the range of characters has been initialised to a large enough size to be safe (which is also the case when you use standard library algorithms such as merge, copy, etc.). If this iterator is an insert_iterator (like back_inserter, ostream_iterator, etc.) then that takes care of growing the output range as we add characters.

Basically, we will use this like this:
```
std::string buffer;
auto output = std::back_inserter( buffer );
chunk.print( output );
```
where ```chunk``` is any type of NDItk component (a table, record, subrecord or a collection of records like the metadata). For the record: the buffer need not be a string per se, you can also use this to directly print to a file or a vector<char> if you wanted to.

ENDFtk and ACEtk use the same style of interface, hence us using it here again.

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+      double awr = 233.0248;
+      double weight = 235.043937521619;
+      double temperature = 2.53e-8;
+      double dilution = 1e+10;
+      multigroup::Structure structure( { 20., 18., 16., 14., 10., 5, 1, 1e-11 } );
+      multigroup::FluxWeights weights( { 0.1, 0.2, 0.25, 0.05, 0.15, 0.04, 0.06 } );
+      multigroup::ReactionCrossSections xs( { { 2, 0.0, { 10., 20., 30., 40., 50., 60., 70. } },
+                                              { 16, 1.1234567, { 1., 2., 3., 4., 5., 6., 7. } } } );
+
+      MultigroupTable chunk( std::move( zaid ), std::move( name ), std::move( source ),
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:38]: This has been a running joke for 6 years now, where the variable name in the unit tests is always ```chunk```. It started with the unit tests in ENDFtk. They are the same in all NJOY libraries.

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+            "be tested" ) {
+
+        verifyChunk( chunk );
+      } // THEN
+
+      THEN( "the record can be printed" ) {
+
+        std::string buffer;
+        auto output = std::back_inserter( buffer );
+        chunk.print( output );
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:52]: Yes, the back_inserter will write to the end of the string (you also have front_inserter. although in this case that would make no sense). It also ensures memory allocation is done correctly. This back_inserter can point to a file, a string, a vector of char, etc.

Uisng a string in the test avoids us creating temporary files on disk - and makes it easier to verify the output by using a string comparison. This is done in ENDFtk and ACEtk as well.

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+      } // THEN
+    } // WHEN
+
+    WHEN( "the data is read using iterators" ) {
+
+      auto iter = record.begin();
+      auto end = record.end();
+
+      MultigroupTable chunk;
+      chunk.read( iter, end );
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:64]: Indeed. The entire content of a file would be read into a string, after which we read over the characters between the iter and end iterator. The read function defined on each table, record and subrecord will only read what it needs (the iter iterator will always point to the first element not read yet) or until the end iterator is encountered.

In this case, the MultigroupTable will start reading a non-whitespace separated string and try to interpret it as a keyword. It then dispatches the reading of the record to the record object storing the information (or to the metadata collection if it is a metadata record).

File: src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+      std::string zaid = "92235.711nm";
+      std::string name = "mendf71x";
+      std::string source = "12/22/2011";
+      std::string process = "08/07/2013";
+      double awr = 233.0248;
+      double weight = 235.043937521619;
+      double temperature = 2.53e-8;
+      double dilution = 1e+10;
+      multigroup::Structure structure( { 20., 18., 16., 14., 10., 5, 1, 1e-11 } );
+      multigroup::FluxWeights weights( { 0.1, 0.2, 0.25, 0.05, 0.15, 0.04, 0.06 } );
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:34]: I did the following: 

https://github.com/njoy/NDItk/blob/8af79148519896510f9b96af03f0a2b51a35d924/src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp#L33
https://github.com/njoy/NDItk/blob/8af79148519896510f9b96af03f0a2b51a35d924/src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp#L134

[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/test/MultigroupTable.test.cpp:34]: These are 9 and 13 digits after the decimal point and they do get printed in the output properly. This is in fix/review by the way.

File: src/NDItk/base/IntegerListRecord.hpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+namespace njoy {
+namespace NDItk {
+namespace base {
+
+/**
+ *  @brief An NDI base record with a list of integer values
+ */
+class IntegerListRecord : protected ListRecord< IntegerListRecord, int > {
+
+  friend class ListRecord< IntegerListRecord, int >;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/IntegerListRecord.hpp:22]: This is inheritance but using what is referred to as a CRTP (Curiously Recursive Template Pattern) in which the base class (```ListRecord< Derived, Type >``` with ```Derived``` = ```IntegerListRecord``` and ```Type```= ```int```) refers to the derived class (in this case ```IntegerListRecord```) for its implementation (this is often referred to as static polymorphism). Where in normal inheritance (i.e. dynamic polymorphism) we would use virtual function to do this, a CRTP does not require virtual functions by referring to the derived class directly. This does away with the additional cost associated to virtual functions.

We have been using CRTPs to do most of our interface inheritance for quite some time, and it is a technique that is getting used more and more often in the development of the standard C++ library (for example: std::ranges::view_interface in C++20 ranges is a CRTP).

The friend declaration is required here to get access to the protected methods on the base class.

For more information on CRTPs: https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern

File: src/NDItk/base/ListRecord.hpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+public:
+
+  using Record::keyword;
+
+  /**
+   *  @brief Return the record values
+   */
+  auto values() const {
+
+    return this->values_ | njoy::tools::std20::views::all;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/ListRecord.hpp:64]: This is a C++20 addition. While we use C++17, this was added in the C++20 extension from njoy::tools (it replaces range-v3). The | operator (or pipe operator) is shorthand for:
```
return njoy::tools::std20::views::all_view( this->values_ );
```

While in this case it offers little advantage, there are cases where it adds readability:
```
auto f = [] ( auto&& value ) { /* do something */ };
return this->values_ | njoy::tools::std20::views::drop( 5 ) | njoy::tools::std20::views::transform( f );
```
which means: "drop the first 5 elements from this->values_ and then apply the transformation f to it". This is equivalent to:
```
auto f = [] ( auto&& value ) { /* do something */ };
return njoy::tools::std20::views::transform_view( njoy::tools::std20::views::drop_view( this->values_, 5 ), f );
```

File: src/NDItk/base/RealListRecord/src/write.hpp [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+  std::ostringstream buffer;
+
+  auto full = this->data().size() / 5;
+  auto partial = this->data().size() - full * 5;
+  auto x = this->data().begin();
+
+  while ( full-- ) {
+
+    buffer << "    " << x[0];
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/RealListRecord/src/write.hpp:19]: Yes, this is where precision went missing. It was added again in fix/review.

File: src/NDItk/multigroup/Metadata/test/Metadata.test.cpp [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+        CHECK( buffer == record );
+      } // THEN
+    } // WHEN
+
+    WHEN( "the data is read using iterators" ) {
+
+      auto iter = record.begin();
+      auto end = record.end();
+
+      auto readKey = [] ( auto&& iter, auto&& end ) {
```
Comment (tgsaller) [inline src/NDItk/multigroup/Metadata/test/Metadata.test.cpp:65]: I don't follow the "auto readKey = []" part.

Comment (tgsaller) [reply src/NDItk/multigroup/Metadata/test/Metadata.test.cpp:65]: Mostly the [] part.

File: src/NDItk/multigroup/ReactionCrossSections/src/read.hpp [hunk 18]
```diff
@@ -0,0 +1,17 @@
+  /**
+   *  @brief Read the record data
+   *
+   *  @param[in] iter   the current position in the input
```
Comment (tgsaller) [inline src/NDItk/multigroup/ReactionCrossSections/src/read.hpp:4]: Can you list the other inputs to read?

File: src/NDItk/multigroup/ReactionCrossSections/src/verify.hpp [hunk 19]
```diff
@@ -0,0 +1,31 @@
+/**
+ *  @brief Verify the groups structure
```
Comment (tgsaller) [inline src/NDItk/multigroup/ReactionCrossSections/src/verify.hpp:2]: These comments are for the group boundaries function.

File: src/NDItk/multigroup/ReactionCrossSections/test/ReactionCrossSections.test.cpp [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+
+std::string chunkWithInsufficientNumberReactions() {
+
+  return "sig_reac\n";
+}
+
+std::string chunkWithInsufficientNumberCrossSectionValues() {
+
+  return "sig_reac\n"
+         "    2 0\n"
```
Comment (tgsaller) [inline src/NDItk/multigroup/ReactionCrossSections/test/ReactionCrossSections.test.cpp:206]: Can you add some cross sections to one of these reactions? So it's like 2 groups for one, and 0 groups for the other.

File: src/NDItk/multigroup/Structure/src/read.hpp [hunk 21]
```diff
@@ -0,0 +1,16 @@
+/**
+ *  @brief Read the record data
+ *
+ *  This function overrides the base record read() function to enable
+ *  verification of the read data.
+ *
+ *  @param[in,out] iter     an iterator to the current position in the input
+ *  @param[in,out] end      an iterator to the end of the input
+ *  @param[in,out] groups   the number of groups to be read
```
Comment (tgsaller) [inline src/NDItk/multigroup/Structure/src/read.hpp:9]: Number of groups + 1

File: src/NDItk/multigroup/Structure/test/Structure.test.cpp [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+        CHECK_THROWS( Structure( std::move( empty ) ) );
+        CHECK_THROWS( Structure( std::move( one ) ) );
+      } // THEN
+    } // WHEN
+
+    WHEN( "the boundary values are not in descending order" ) {
+
+      THEN( "an exception is thrown" ) {
+
+        std::vector< double > wrong = { 1., 20. };
```
Comment (tgsaller) [inline src/NDItk/multigroup/Structure/test/Structure.test.cpp:92]: Can you make this something like { 3., 1., 20. }?

Comment (tgsaller) [reply src/NDItk/multigroup/Structure/test/Structure.test.cpp:92]: Or both.

File: src/NDItk/base/SubListRecord/src/ctor.hpp [hunk 23]
```diff
@@ -0,0 +1,60 @@
+SubListRecord() = default;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/SubListRecord/src/ctor.hpp:1]: This tells the compiler to just go with the default generated constructor (i.e. a constructor for SubListRecord that has no arguments and that just calls the default constructors for the underlying fields). In this case, this is equivalent to:
```
SubListRecord() : values_(), begin_(), end_(), length_() {}
```

This does not work when the fields do not have default constructors themselves (the compiler will warn about this).

File: src/NDItk/base/SubListRecord/src/ctor.hpp [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+
+  *this = base;
+}
+
+SubListRecord( SubListRecord&& base ) {
+
+  *this = std::move( base );
+}
+
+SubListRecord( const Iterator& begin, const Iterator& end ) :
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/SubListRecord/src/ctor.hpp:13]: No. A SubListRecord refers to a subrecord of another record (it lifts a view of values out of the other record). When this constructor is used, the data is not owned by the SubListRecord, we just refer to it.

[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/SubListRecord/src/ctor.hpp:13]: The record from which the SubListRecord came takes caring of reading teh data. You'll see this in the ReactionCrossSections record where generateBlocks() will create the subrecords.

File: src/NDItk/base/SubListRecord/src/ctor.hpp [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+
+SubListRecord( std::vector< Type > values ) :
+  values_( std::move( values ) ) {
+
+  this->begin_ = this->values_->begin();
+  this->end_ = this->values_->end();
+  this->length_ = this->values_->size();
+}
+
+SubListRecord& operator=( const SubListRecord& right ) {
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/SubListRecord/src/ctor.hpp:26]: That is a personal convention: it is on the right side of the operator:
```left = right```

File: src/NDItk/multigroup/CrossSection/src/write.hpp [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ *  @param[in] iter   the current position in the output
+ */
+template< typename OutputIterator >
+void write( OutputIterator& iter ) const {
+
+  std::ostringstream buffer;
+
+  auto x = this->begin();
+  buffer << "    " << x[0] << ' ' << std::setprecision( 15 ) << x[1] << '\n';
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/CrossSection/src/write.hpp:14]: It should be an integer in this case (it is the MT number), although a static_cast<int>( std::round(  ) ) may be in order here just in case. I'll modify fix/review.

File: src/NDItk/multigroup/FluxWeights.hpp [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+#include "NDItk/base/RealListRecord.hpp"
+
+namespace njoy {
+namespace NDItk {
+namespace multigroup {
+
+/**
+ *  @brief A flux weight record
+ */
+class FluxWeights : protected base::RealListRecord {
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/FluxWeights.hpp:17]: This is mainly to reuse the interface and potentially redefine the function names. C++ is a strongly typed language. While we could use RealListRecord all over the place, a GroupStructure and FluxWeights record are not the same while RealListRecord are the same even if they contain different types of records.

File: src/NDItk/multigroup/Metadata/test/Metadata.test.cpp (hunk 17 above)
Comment (whaeck) [reply src/NDItk/multigroup/Metadata/test/Metadata.test.cpp:65]: This is a lambda. It allows us to define a function in the code. the ```[]``` is what we refer to as the capture for the lambda (i.e. things the lanbda needs to do its job but that you cannot give in the lambda's arguments). In this case, ```[]``` means that nothing needs to be captured.

For example:
```
int id = 4;
auto compare = [id] ( auto&& value ) { return value == id; };
```
In this case the value of ```id``` gets captured by value in the lambda.

File: src/NDItk/multigroup/CrossSection/src/write.hpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (tgsaller) [reply src/NDItk/multigroup/CrossSection/src/write.hpp:14]: Good idea!

File: src/NDItk/multigroup/ReactionCrossSections/src/read.hpp (hunk 18 above)
Comment (whaeck) [reply src/NDItk/multigroup/ReactionCrossSections/src/read.hpp:4]: This has been fixed in the next PR: #3

File: src/NDItk/multigroup/ReactionCrossSections/src/verify.hpp (hunk 19 above)
Comment (whaeck) [reply src/NDItk/multigroup/ReactionCrossSections/src/verify.hpp:2]: Gotta love copy/paste. Fixed in fix/review.

File: src/NDItk/multigroup/Structure/src/read.hpp (hunk 21 above)
Comment (whaeck) [reply src/NDItk/multigroup/Structure/src/read.hpp:9]: Yes. I fixed this in fix/review some time ago.

File: src/NDItk/multigroup/Structure/test/Structure.test.cpp (hunk 22 above)
Comment (whaeck) [reply src/NDItk/multigroup/Structure/test/Structure.test.cpp:92]: Done. Changed in fix/review

File: src/NDItk/multigroup/CrossSection/src/write.hpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/CrossSection/src/write.hpp:14]: Done. Added to fix/review

File: src/NDItk/multigroup/CrossSection/test/CrossSection.test.cpp [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+
+  GIVEN( "valid data for a CrossSection instance" ) {
+
+    std::string record = chunk();
+
+    WHEN( "the data is given explicitly" ) {
+
+      int reaction = 2;
+      double qvalue = 0.0;
+      std::vector< double > values = { 10., 20., 30., 40., 50., 60., 70. };
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/CrossSection/test/CrossSection.test.cpp:30]: Since we override the print function, yes, this is a good idea.

Changes made to fix/review. I also added additional digits there for the RealListRecord test.

File: src/NDItk/base/Record.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/Record.hpp:40]: One could use sfinae to do this, there's an actual std::output_iterator< T > in C++20 to test whether or not some type T is an output iterator. In this context however I don't think it matters. No user is going to directly use the print function.

File: src/NDItk/base/IntegerListRecord/src/write.hpp [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+ */
+template< typename OutputIterator >
+void write( OutputIterator& iter ) const {
+
+  std::ostringstream buffer;
+
+  auto full = this->data().size() / 5;
+  auto partial = this->data().size() - full * 5;
+  auto x = this->data().begin();
+
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/IntegerListRecord/src/write.hpp:16]: Not sure, but magic numbers are to be avoided anyway so why not change it? Other than global consts I don't see an easy way to do this though.

The same would go for the output precision too.

File: src/NDItk/base/IntegerListRecord/src/write.hpp [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ *  @param[in] iter   the current position in the output
+ */
+template< typename OutputIterator >
+void write( OutputIterator& iter ) const {
+
+  std::ostringstream buffer;
+
+  auto full = this->data().size() / 5;
+  auto partial = this->data().size() - full * 5;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/IntegerListRecord/src/write.hpp:14]: Yes, you are correct. I'll change it.

File: src/NDItk/base/Record.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (joewpeterson) [reply src/NDItk/base/Record.hpp:40]: what concerns me is making this a long lasting program. In the next 50 years is any developer going to unwittingly pass in a standard iterator to this template and cause the compiler to write a really unsafe function? The compiler is not going to complain - it is perfectly valid to pass this in. The program may look like it works in certain situations. It would be really hard to debug this if a small accident occurred in development through developer ignorance or mistake.

I get where you're coming from and that these exist a lot - memcpy - for one. But this isn't a well known function. This exists just here in this program. Perhaps the name of the variable is enough here - OutputIterator - to document and note that this was only supposed to be an output iterator for safe use. But other than this implicit note, there is not much documenting how careful one must be with this template. It just makes me nervous to see this, you know?

[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/Record.hpp:40]: This style is very similar to the C++ standard library, so any seasoned C++ programmer should recognise this immediately (and understand the underlying limitations). Look here for an example: https://en.cppreference.com/w/cpp/algorithm/copy or https://en.cppreference.com/w/cpp/algorithm/transform

Now, on the reverse side of the argument I do understand. I have been guilty of forgetting to allocate a vector and just use the begin iterator to an empty vector in a STL algorithm. As a result, you always do things like this:
https://github.com/njoy/scion/blob/d828ef47aa0d60daece9cdd5ff486d76aa6c8882/src/scion/unionisation/unionise.hpp#L27 if you are not using something like a back_inserter.

[the remark this answers is not in the document]
Comment (joewpeterson) [reply src/NDItk/base/Record.hpp:40]: Alright, sounds good. If you are comfortable with this, I'm fine. Its good to have the discussion here. Hopefully, if an error does occur down the line, this conversation with hasten its resolution.

File: src/NDItk/MultigroupTable/src/read.hpp [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+
+    if ( this->metadata_.isMetadataKey( keyword ) ) {
+
+      this->metadata_.read( keyword, iter, end );
+    }
+    else if ( keyword == this->structure_.keyword() ) {
+
+      if ( this->metadata_.numberGroups().has_value() ) {
+
+        readRecord( this->structure_, iter, end,
```
[the remark this answers is not in the document]
Comment (tgsaller) [reply src/NDItk/MultigroupTable/src/read.hpp:22]: outgoingGroupStructure only applies for secondary particles? And is ( 0 ) index 0, or saying particle type 0 (gamma)?

File: src/NDItk/base/ListRecord.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/base/ListRecord.hpp:64]: While this is indeed a c++20 extension, it was implemented in the tools library using c++17 standard programming. The interface is in line with the c++20 specs. I have tested this even on intel-classic compilers and it works. You should not need a C++20 compatible compiler, you need a c++17 compatible one. Once we can actually switch standards, we can do away with all this.

One of the things we have to do soon is extend the testing capabilities we have to test more OS types and compilers. I plan on talking to the MCNP team about that.

**Review by tgsaller — APPROVED** [review]
Looks good!
